dawn/docs/parallel_model.rst

```diff
@@ -0,0 +1,251 @@
+   output fields we duplicate the writes which can have 2 side-effects
+   if neighboring blocks are out of sync: Block 1 writes in its extended
+   domain (line 2), block 2 writes (line 2) (same value), then reads
+   offset (line 3), then writes in the center (line 4). Now Block 1
+   reads (line 3) and gets the wrong value. This means that if we want
+   to solve the problem we need a global sync in that case (meaning
+   breaking the kernel call).
+
+.. code:: c++
+
+   1 vertical_region(k_start, k_end){
+   2   one = 1
+   3   two = one[i - 1]
+   4   one = 2
+   5 }
+
+translates to
+
+::
+
+   MultiStage_0 [parallel]
+       {
+         Stage_0
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 1;
+           }
+           Extents: [(-1,0),(0,0),(0,0)]
+         }
+         Stage_1
+         {
+           Do_0 { Start : End }
+           {
+             two[<no_horizontal_offset>,0] = one[-1,0,0];
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+         Stage_2
+         {
+           Do_0 { Start : End }
+           {
+             one[<no_horizontal_offset>,0] = int_type 2;
+           }
+           Extents: [<no_horizontal_extent>,(0,0)]
+         }
+       }
+
+--------------
+
+Compute Domain
+~~~~~~~~~~~~~~
+
+If the vertical region does not specify any horizontal extent, dawn
+ensures that the compute-domain of statements that others depend on is
+extended to fill all the intermediate fields as required. If a
+horizontal compute-domain is specified, it is never extended.
+
+Internal Contract of Data Structures
+------------------------------------
+
+StencilInstantiation
+~~~~~~~~~~~~~~~~~~~~
+
+StencilInstantiations (SI) are user defined and are never changed during
+any optimization. Each SI will be globally split from every other one
+and there is no optimization across SIs. For every SI, a callable API
+will be generated.
+
+Stencil
+~~~~~~~
+
+Stencils serve no purpose right now
+
+--------------
+
+   It might be useful to remove them altogether since even if they
+   existed, no concept can't be captured by the other structures
+
+--------------
+
+Multistage
+~~~~~~~~~~
+
+A Multistage (MS) is the equivalent of a global synchronization between
+all processes on the full domain. each Multistage can only have one loop
+order
+
+-  in the CUDA back end they are the equivalent of a kernel call
+-  in the GridTools back end they are the equivalent of a multistage
+-  in the naive back end every storage will be synchronized on the host
+   at the beginning of each multistage
+
+--------------
+
+   Can we discuss why we still have the loop order on this level? Would
+   it potentially be interesting to promote this to the Do level and
+   change the fusion strategy if we use the GridTools back end? Or do we
+   know that we are losing performance in the CUDA back end if we mix
+   forward and backward loops versus new kernel calls?
```

Comment:
I don't think we risk loosing performance if we remove the loop order from the multistage, or, at least, I couldn't come up with an example for which we would miss possible optimizations with such parallel model. Just having the loop order on the do-methods should do fine. (I assume the various reordering/fusing/splitting algorithms are correct and smart)
An example to motivate is:
```
1 one = 1
2 zero = 0
3 two = one[k - 1]       // forward
4 three = one[k + 1] 	   // backwards
5 one = 2
6 four = zero[k - 1]     // forward
```
here we would need a multistage split between lines 4 and 5. But this shouldn't prevent line 6 to be "put together" with line 3 (loops fused because of same order), as it has no dependency on lines 4 and 5.